### DISCUSSION TITLE
How to access navigation events?

### DISCUSSION THREAD

**Discussion by mtr1990** [igordanchenko/yet-another-react-lightbox#31] [Q&A] [answered]
How can we access `onNext` events when 'next button' is clicked and `onPrev` when ' prev button' is clicked?

**Comment by igordanchenko** [comment, marked as answer]
@mtr1990, can you describe what you are trying to accomplish? 

Have you considered using `view` event instead?

```jsx
<Lightbox on={{ view: (index) => {} }} />
```

**Reply by mtr1990** [reply]
Hi @igordanchenko ,

I knew about this event. And I got the current index through this event.

But I am referring to other event like click next/prev button

Example 

```js
<Lightbox on={{
 onNext: () => console.log('next image')
 onPrev: () => console.log('prev image')
 }} />
```

**Reply by igordanchenko** [reply]
The lightbox doesn't provide such events because I couldn't come up with a use case for them. That's why I'm wondering what you are trying to accomplish with `onPrev` / `onNext` callbacks.

Here is how you can detect `prev` /  `next` events inside the `view` event:

```jsx
const currentIndexRef = React.useRef();

<Lightbox
  on={{
    view: (currentIndex) => {
      const previousIndex = currentIndexRef.current;
      currentIndexRef.current = currentIndex;

      if (previousIndex !== undefined && previousIndex !== currentIndex) {
        //...
      }
    }
  }}
  //...
/>
```

Does this help?

**Reply by minimal-ui-kit** [reply]
```js
const SlideWrapper = ({ children }) => {
  const { publish } = useEvents();

  // supported events: "close", "prev", "next"
  return <div onClick={() => publish("close")}>{children}</div>;
};
```

I thought I was referring to this one. Where events are accessible.


:D

**Reply by igordanchenko** [reply]
Well, it's still a mystery to me what you are trying to accomplish, but here is how you can access the internal "prev" / "next" events in the above example:

```jsx
import {
  cleanup,
  useEvents,
  useLightboxState,
  useController,
  ACTION_PREV,
  ACTION_NEXT
} from "yet-another-react-lightbox/core";

const SlideWrapper = ({ slide, children }) => {
  const { publish, subscribe } = useEvents();
  const {
    state: { currentIndex }
  } = useLightboxState();
  const { getLightboxProps } = useController();
  const { slides } = getLightboxProps();

  React.useEffect(
    () =>
      slides[currentIndex] === slide
        ? cleanup(
            subscribe(ACTION_PREV, (count) => {
              console.log("prev");
            }),
            subscribe(ACTION_NEXT, (count) => {
              console.log("next");
            })
          )
        : () => {},
    [subscribe, slides, currentIndex, slide]
  );

  // supported events: "close", "prev", "next"
  return <div onClick={() => publish("close")}>{children}</div>;
};
```

Is this what you are looking for?

DISCLAIMER: The above code snippet was provided as part of the exploratory conversation and is not endorsed for use in production because the internal "prev" and "next" events are probably not what you expect them to be.

**Reply by mtr1990** [reply]
Hi,

Thanks for creating this component and your support. The code you shared works fine. (Expecting you to update in the docs for this.)

---

What I expected more would probably be like:


```js
<Lightbox 
 onNext={() => {}}
 onClose={() => {}}
 onPrev={() => {}}
>
...
</Lightbox>
```

OR

```js
<Lightbox 
on={{
 view: (index) => {},
 next: () => {
 // other action when next event
  },
 prev: () => {
 // other action when prev event
 },
 close: () => {
 // other action when close event
 },
}}
>
  
```

Better for DX. Since these are fixed events that do not change

This is just my personal opinion. You can skip. :)

**Reply by igordanchenko** [reply]
@mtr1990

Well, here is the thing - these _internal events_ are not part of the public API because they are probably not what you expect them to be. Both `prev` and `next` _internal events_ are not events per se but requests for that action. For instance, `prev` can be better interpreted as a "request to transition lightbox to the previous slide" - it's a request, not a guaranteed action or the actual event. For example, if you open a lightbox with `carousel={{ finite: true }}` and hit the `ArrowLeft` button on your keyboard, that would trigger the internal `prev` event because your input device (your keyboard in this case) is trying to navigate to the previous slide. But because the lightbox is in the "finite" mode, it doesn't honor this request, and no actual event takes place. I hope this example clarifies the difference.

I'll be happy to update the public API with an appropriate set of events to improve the DX as long as I'm clear about the intended use of the events. And for that, I need at least one real-life example of how they will actually be used. The last thing I want to do is to build a wrong thing based on incorrect assumptions. Please feel free to open a feature request with a detailed description and example(s).

P.S. There was an issue with the code snippet I provided yesterday - the callbacks were called multiple times per event (`2 * carousel.preload + 1` to be exact). I've updated the above example and added a disclaimer clarifying it's not recommended for use in production.

**Reply by JurajKavka** [reply]
Hi, I'm also looking for such events. I would like to update URL in browser while switching images in Lightbox. Each image should have its own unique URL in browser so when you hit refresh Light box should come up with image id provided in search parameters. This is standard behavior of galleries in media like newspaper portals etc. So to have onNext() onPrev() events (or better overrides) wold be highly appreciated.

**Reply by igordanchenko** [reply]
@JurajKavka it sounds like you are looking for the `on.view` callback. The `on.view` callback gets called every time the slide index changes, and that's the perfect opportunity to update the browser URL. 

Please see the below example and let me know if you have any questions.

<details>
<summary>Example (Next.js 13 app router)</summary>

```tsx
"use client";

import * as React from "react";
import Lightbox from "yet-another-react-lightbox";
import { useRouter, useSearchParams } from "next/navigation";

import { slides } from "@/components/photos";

export default function Example() {
    const router = useRouter();
    const searchParams = useSearchParams();
    const indexSearchParam = searchParams.get("index");

    const [open, setOpen] = React.useState(() => Boolean(indexSearchParam));
    const [index, setIndex] = React.useState(() => Number.parseInt(indexSearchParam || "0", 10));

    const updateIndex = (currentIndex?: number) => {
        setIndex(currentIndex || 0);

        const url = new URL(window.location.href);
        const queryParams = url.searchParams;
        if (currentIndex !== undefined) {
            queryParams.set("index", `${currentIndex}`);
        } else {
            queryParams.delete("index");
        }
        url.search = queryParams.toString();

        router.replace(url.toString(), { scroll: false });
    };

    return (
        <>
            <Lightbox
                slides={slides}
                index={index}
                open={open}
                close={() => {
                    updateIndex();
                    setOpen(false);
                }}
                on={{
                    view: ({ index: currentIndex }) => {
                        updateIndex(currentIndex);
                    },
                }}
            />

            <button type="button" onClick={() => setOpen(true)}>
                Open Lightbox
            </button>
        </>
    );
}
```
</details>

**Reply by JurajKavka** [reply]
Yes, that's great and working example, but I found that back and forward buttons in browser does not work properly. I would like to use `push` not `replace` and If I hit "back" I can read URL from `window.location.href` but I can't call `setIndex` directly because whole Lightbox rerenders and  `updateIndex` is called twice for same URL. Gallery should listen to URL changes and slide images according those changes. Is that possible?

**Reply by igordanchenko** [reply]
Can you point me to a live example of a gallery with such UX?

**Reply by JurajKavka** [reply]
I found one [here](https://diva.aktuality.sk/fotogaleria/WSvCJcu/bez-filtrov-a-nazivo-premene-nasej-znamej-influencerky-neverili-ona-ju-teda-predviedla-v-realite/1/), but lots of big players does not have that feature in their galleries...

**Reply by igordanchenko** [reply]
Thank you for this example. While unconventional, I guess in this case you can store slide index state directly in URL query parameter, and conditionally update it when current index doesn't match the one in the URL (which effectively means the lightbox transitioned to a different slide).

<details>
<summary>Example (Next.js 13 app router)</summary>

```jsx
"use client";

import * as React from "react";
import Lightbox from "yet-another-react-lightbox";
import { useRouter, useSearchParams } from "next/navigation";

import { slides } from "@/components/photos";

export default function Example() {
    const router = useRouter();
    const searchParams = useSearchParams();

    const param = searchParams.get("index");
    const index = param ? Number.parseInt(param, 10) : undefined;

    const updateIndex = (currentIndex?: number) => {
        if (index !== currentIndex) {
            const url = new URL(window.location.href);
            const queryParams = url.searchParams;

            if (currentIndex !== undefined) {
                queryParams.set("index", `${currentIndex}`);
            } else {
                queryParams.delete("index");
            }
            url.search = queryParams.toString();
            
            router.push(url.toString(), { scroll: false });
        }
    };

    return (
        <>
            <Lightbox
                slides={slides}
                index={index}
                open={index !== undefined}
                close={() => updateIndex()}
                on={{ view: ({ index: currentIndex }) => updateIndex(currentIndex) }}
            />

            <button type="button" onClick={() => updateIndex(0)}>
                Open Lightbox
            </button>
        </>
    );
}
```
</details>

**Reply by JurajKavka** [reply]
Hi Thank You for your help! I used that example and my gallery works as expected. I have another question - would be possible to build npm package as CommonJS moule? We are on CommonJS environment and when I start express with SSR service I get an error

```
 require() of ES Module /XYZ/node_modules/yet-another-react-lightbox/dist/index.js from /XYZ/node_modules/XYZ/PhotoGallery/SimplePhotoGallery.js not supported.
Instead change the require of index.js in /XYZ/PhotoGallery/SimplePhotoGallery.js to a dynamic import() which is available in all CommonJS modules.
```

**Reply by igordanchenko** [reply]
> I have another question 

Do you mind opening a separate discussion? This topic is totally unrelated to the original question, and opening a separate discussion would give other users a chance of finding this question in the future.
